### PR TITLE
mpvScripts.autosub: init at 2021-06-29

### LINKED ISSUE #336986: Package request: mpvScripts.autosub
**Project description**

[Automatic subtitle downloading for MPV](https://github.com/davidde/mpv-autosub)

**Metadata**

* homepage URL: https://github.com/davidde/mpv-autosub
* source URL: https://github.com/davidde/mpv-autosub
* license: MIT
* platforms: unix, linux, darwin

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by octvs**
Add `mpvScripts.autosub`, resolves #336986.

Took the discussion over at the linked issue, and other mpvScripts as basis.
~~Couple of questions to whom might review this:~~

- ~~There are usage of `passthru.updateScript` on other scripts, what is the best
practice for this?~~
- ~~In a similar manner I saw uses of `lib.getExe` for dependencies on
`substituteInPlace` like [here][1]. What is the point of using it rather than a
direct path?~~
  - ~~The latter seems more safe (or exact) _to me_, therefore I opted for using
  it. But I would like to know if there is a best practice and/or the logic
  behind using `lib.getExe`.~~

Cheers!

_Credits to @jcaesar for supplying an initial version at the discussion_

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

[1]: https://github.com/NixOS/nixpkgs/blob/70548b93f05d02cb6e90750272e3b6531ee26d7d/pkgs/applications/video/mpv/scripts/videoclip.nix#L28-L29

### CODE REVIEW

**Comment by octvs** [comment]
Currently failing test is due to my entry from `lib.maintainers` missing, which
I expected to be present before this one with [this PR] already merging.

[this PR]: https://github.com/NixOS/nixpkgs/pull/354823

**Comment by AndersonTorres** [comment]
> What is the point of using it rather than a
direct path?

Encapsulation.
lib.getExe will pick the executable from `outputBin`. And `outputBin` can assume different paths according to the derivation, especially in split outputs.

**Comment by octvs** [comment]
> Encapsulation.
> lib.getExe will pick the executable from `outputBin`. And `outputBin` can assume different paths according to the derivation, especially in split outputs.

Thanks for the kind explanation, I changed it to use `lib.getExe'`.

**Comment by octvs** [comment]
I've found the answer to the other question [here][1].


[1]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#automatic-package-updates

File: pkgs/applications/video/mpv/scripts/autosub.nix
Comment (AndersonTorres) [inline pkgs/applications/video/mpv/scripts/autosub.nix:6]:
```suggestion
  python3Packages,
```

This is a bit of dark magic to explain, just believe me here: `python3.pkgs` **is not** equivalent to `python3Packages`.

Nixpkgs is crafted so that the typical `mkDerivation` is transparent to cross-compilation.
However, this dark magic fails in expressions like `python3.pkgs`, while working in `python3Packages`.

File: pkgs/applications/video/mpv/scripts/autosub.nix
Comment (AndersonTorres) [inline pkgs/applications/video/mpv/scripts/autosub.nix:25]:
```suggestion
      "local subliminal = '${lib.getExe' python3Packages.subliminal "subliminal"}'"
```

File: pkgs/applications/video/mpv/scripts/autosub.nix
Comment (AndersonTorres) [inline pkgs/applications/video/mpv/scripts/autosub.nix:20]: Put this right before `meta`.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by octvs** [comment]
Alright, suggestions are applied, thanks for the feedback & explanation!

**Review by AndersonTorres — APPROVED** [review]

**Comment by jcaesar** [comment]
Depending on your progress with #354823, you could also cherry-pick the [maintainer add](https://github.com/NixOS/nixpkgs/pull/354823/commits/0f27402c0944d32327479401a3b0ad3aca68d248) commit onto here. It won't cause merge conflicts, and if you rebase either PR after the other is merged, the duplicate commit will silently disappear.

**Comment by octvs** [comment]
> Depending on your progress with #354823, you could also cherry-pick the [maintainer add](https://github.com/NixOS/nixpkgs/pull/354823/commits/0f27402c0944d32327479401a3b0ad3aca68d248) commit onto here. It won't cause merge conflicts, and if you rebase either PR after the other is merged, the duplicate commit will silently disappear.

Wow, that is great to know. Finally I get to use the `cherry-pick`.

Thank you for the kind tip!

**Review by nicoonoclaste — APPROVED** [review]

**Pull request merged**
